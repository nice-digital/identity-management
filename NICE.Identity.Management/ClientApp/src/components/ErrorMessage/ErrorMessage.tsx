import React from "react";

import { Alert } from "@nice-digital/nds-alert";

type ErrorMessageProps = {
	error: Error;
	children?: React.ReactNode;
};

export const ErrorMessage = (props: ErrorMessageProps): React.ReactElement => (
	<Alert type="error">
		{props.children ? (
			props.children
		) : (
			<>
				<h2>Whoops</h2>
				<p>Sorry, there&apos;s been an error!</p>
			</>
		)}

		{process.env.NODE_ENV === "development" && (
			<dl>
				<dt>Message:</dt>
				<dd>{props.error.message}</dd>
				<dt>Stack:</dt>
				<dd>{props.error.stack}</dd>
			</dl>
		)}
	</Alert>
);
