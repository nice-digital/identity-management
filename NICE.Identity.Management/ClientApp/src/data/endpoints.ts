const API_BASE_URL = process.env.REACT_APP_API_BASE_URL
	? process.env.REACT_APP_API_BASE_URL
	: "#{REACT_APP_API_BASE_URL}";

const APP_BASE_URL = `${API_BASE_URL}` == 'http://api:8090' ? `${API_BASE_URL}` : API_BASE_URL.replace("/api", '');

export const Endpoints = {
	identityManagementUser: `${APP_BASE_URL}/account/status`, // accesses the account controller to get the status endpoint which returns JSON data containing the username.
	job: (jobId: string | number): string => `${API_BASE_URL}/jobs/${jobId}`,
	jobs: `${API_BASE_URL}/jobs`,
	organisation: (organisationId: string | number): string => `${API_BASE_URL}/organisations/${organisationId}`,
	organisationsList: `${API_BASE_URL}/organisations`,
	organisationsListSearch: (searchQuery: string): string => `${API_BASE_URL}/organisations?q=${searchQuery}`,
	service: (serviceId: string | number): string => `${API_BASE_URL}/services/${serviceId}`,
	servicesList: `${API_BASE_URL}/services`,
	user: (userId: string | number): string => `${API_BASE_URL}/users/${userId}`,
	usersList: `${API_BASE_URL}/users`,
	usersListSearch: (searchQuery: string): string => `${API_BASE_URL}/users?q=${searchQuery}`,
	usersByOrganisation: (organisationId: string | number): string => `${API_BASE_URL}/users/usersbyorganisation/${organisationId}`,
	usersAndJobIdsByOrganisation: (organisationId: string | number): string => `${API_BASE_URL}/users/usersandjobidsbyorganisation/${organisationId}`,	
	userRolesByWebsite: (userId: string | number, websiteId: string | number): string => `${API_BASE_URL}/users/${userId}/rolesbywebsite/${websiteId}`,
	usersAndRolesByWebsite: (websiteId: string | number): string =>	`${API_BASE_URL}/websites/${websiteId}/usersandrolesbywebsite`,
	verificationEmail: `${API_BASE_URL}/VerificationEmail/VerificationEmail`,
	website: (websiteId: string | number): string => `${API_BASE_URL}/websites/${websiteId}`,		
	websitesList: `${API_BASE_URL}/websites`,
};
