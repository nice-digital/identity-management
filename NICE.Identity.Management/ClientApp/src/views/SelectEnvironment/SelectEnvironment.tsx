import React, { Component } from "react";
import { RouteComponentProps, Link } from "react-router-dom";

import { Endpoints } from "../../data/endpoints";
import { fetchData } from "../../helpers/fetchData";
import { isDataError } from "../../helpers/isDataError";
import { UserType, ServiceType, WebsiteType } from "../../models/types";
import { PageHeader } from "@nice-digital/nds-page-header";
import { Grid, GridItem } from "@nice-digital/nds-grid";
import { StackedNav, StackedNavLink } from "@nice-digital/nds-stacked-nav";
import { ErrorMessage } from "../../components/ErrorMessage/ErrorMessage";
import { Breadcrumbs, Breadcrumb } from "@nice-digital/nds-breadcrumbs";

type TParams = { id: string; serviceId: string };

type SelectEnvironmentProps = Record<string, unknown> &
	RouteComponentProps<TParams>;

type SelectEnvironmentState = {
	user: UserType;
	service: ServiceType;
	websites: Array<WebsiteType>;
	error?: Error;
	isLoading: boolean;
};

export class SelectEnvironment extends Component<
	SelectEnvironmentProps,
	SelectEnvironmentState
> {
	constructor(props: SelectEnvironmentProps) {
		super(props);
		this.state = {
			user: {} as UserType,
			service: {} as ServiceType,
			websites: [],
			isLoading: true,
		};
		document.title = "NICE Accounts - Select environment";
	}

	async componentDidMount(): Promise<void> {
		this.setState({ isLoading: true });

		const user = await fetchData(Endpoints.user(this.props.match.params.id));
		const service = await fetchData(
			Endpoints.service(this.props.match.params.serviceId),
		);

		if (isDataError(user)) {
			this.setState({ error: user });
		}
		if (isDataError(service)) {
			this.setState({ error: service });
		}

		const websites = service.websites;

		this.setState({ user, service, websites, isLoading: false });
	}

	render(): JSX.Element {
		const { user, service, websites, error, isLoading } = this.state;
		const { params, url } = this.props.match;
		const { id } = params;

		let nameBreadcrumb = `${user.firstName} ${user.lastName}`;

		if (isLoading) {
			nameBreadcrumb = "Loading user details";
		}
		if (error) {
			nameBreadcrumb = "Error";
		}

		return (
			<>
				<Breadcrumbs>
					<Breadcrumb to="/users" elementType={Link}>
						Users
					</Breadcrumb>
					<Breadcrumb to={`/users/${id}`} elementType={Link}>
						{nameBreadcrumb}
					</Breadcrumb>
					<Breadcrumb to={`/users/${id}/services`} elementType={Link}>
						Select service
					</Breadcrumb>
					<Breadcrumb>Select environment</Breadcrumb>
				</Breadcrumbs>

				{!error ? (
					<>
						<PageHeader
							preheading="Select environment for"
							heading={isLoading ? "Service name" : service.name}
						/>

						<Grid>
							<GridItem cols={8}>
								{isLoading ? (
									<p>Loading...</p>
								) : (
									<StackedNav aria-label="Environments">
										{websites.map((website) => {
											return (
												<StackedNavLink
													destination={`${url}/${website.id}/roles`}
													elementType={Link}
													key={website.id}
												>
													{website.environment.name}
												</StackedNavLink>
											);
										})}
									</StackedNav>
								)}
							</GridItem>
						</Grid>
					</>
				) : (
					<>
						<PageHeader heading="Error" />
						<ErrorMessage error={error}></ErrorMessage>
					</>
				)}
			</>
		);
	}
}
