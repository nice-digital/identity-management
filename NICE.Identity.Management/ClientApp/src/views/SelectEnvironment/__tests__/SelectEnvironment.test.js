import React from "react";
import { mount, shallow } from "enzyme";
import { MemoryRouter } from "react-router";
import toJson from "enzyme-to-json";

import { nextTick } from "../../../utils/nextTick";
import singleUser from "./singleUser.json";
import singleService from "./singleService.json";
import { SelectEnvironment } from "../SelectEnvironment";

import { Endpoints } from "../../../data/endpoints";
import { ErrorMessage } from "../../../components/ErrorMessage/ErrorMessage";

describe("SelectEnvironment", () => {
	const match = {
		params: { id: 1, serviceId: 1 },
		isExact: true,
		path: "",
		url: "",
	};

	const consoleErrorReset = console.error;

	beforeEach(() => {
		fetch.resetMocks();
		console.error = consoleErrorReset;
	});	

	it("should show loading message before data has been loaded", () => {
		fetch.mockResponseOnce(JSON.stringify(singleUser));
		fetch.mockResponseOnce(JSON.stringify(singleService));
		const wrapper = shallow(<SelectEnvironment match={match} />);
		expect(wrapper.find("p").text()).toEqual("Loading...");
	});

	it("should call fetch during componentDidMount", async () => {
		fetch.mockResponseOnce(JSON.stringify(singleUser));
		fetch.mockResponseOnce(JSON.stringify(singleService));
		const wrapper = mount(<MemoryRouter><SelectEnvironment match={match} /></MemoryRouter>);
		const spy = jest.spyOn(wrapper.instance(), "componentDidMount");
		wrapper.instance().componentDidMount();
		await nextTick();
		wrapper.update();
		expect(spy).toHaveBeenCalled();
		expect(fetch.mock.calls.length).toEqual(2);
		expect(fetch.mock.calls[0][0]).toEqual(Endpoints.user(match.params.id));
		expect(fetch.mock.calls[1][0]).toEqual(Endpoints.service(match.params.serviceId));
		spy.mockClear();
	});

	it("should match the snapshot after data has been loaded", async () => {
		fetch.mockResponseOnce(JSON.stringify(singleUser));
		fetch.mockResponseOnce(JSON.stringify(singleService));
		const wrapper = shallow(<SelectEnvironment match={match} />);
		await nextTick();
		wrapper.update();
		expect(toJson(wrapper, { noKey: true, mode: "deep" })).toMatchSnapshot();
	});

	it("should show error message when user fetchData function returns 500 error", async () => {
		console.error = jest.fn();		
		fetch.mockRejectOnce(new Error("500 Internal Server Error"));
		fetch.mockResponseOnce(JSON.stringify(singleService));
		const wrapper = mount(
			<MemoryRouter>
				<SelectEnvironment match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		expect(wrapper.find(ErrorMessage).exists()).toBe(true);
	});

	it("should show error message when user fetchData function returns 401 error", async () => {
		console.error = jest.fn();		
		fetch.mockResponseOnce(JSON.stringify({}), { status: 401 });
		fetch.mockResponseOnce(JSON.stringify(singleService));
		const wrapper = mount(
			<MemoryRouter>
				<SelectEnvironment match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		expect(wrapper.find(ErrorMessage).exists()).toBe(true);
	});

	it("should show error message when services fetchData function returns 500 error", async () => {
		console.error = jest.fn();		
		fetch.mockResponseOnce(JSON.stringify(singleUser));
		fetch.mockRejectOnce(new Error("500 Internal Server Error"));
		const wrapper = mount(
			<MemoryRouter>
				<SelectEnvironment match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		expect(wrapper.find(ErrorMessage).exists()).toBe(true);
	});

	it("should show error message when services fetchData function returns 401 error", async () => {
		console.error = jest.fn();		
		fetch.mockResponseOnce(JSON.stringify(singleUser));
		fetch.mockResponseOnce(JSON.stringify({}), { status: 401 });
		const wrapper = mount(
			<MemoryRouter>
				<SelectEnvironment match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		expect(wrapper.find(ErrorMessage).exists()).toBe(true);
	});
});
