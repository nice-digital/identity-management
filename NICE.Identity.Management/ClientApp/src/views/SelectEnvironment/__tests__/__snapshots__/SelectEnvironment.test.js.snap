// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectEnvironment should match the snapshot after data has been loaded 1`] = `
<Fragment>
  <Breadcrumbs>
    <Breadcrumb
      data-qa-sel="breadcrumb-administration-link"
      elementType={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "propTypes": Object {
            "innerRef": [Function],
            "onClick": [Function],
            "replace": [Function],
            "target": [Function],
            "to": [Function],
          },
          "render": [Function],
        }
      }
      to="/"
    >
      Administration
    </Breadcrumb>
    <Breadcrumb
      data-qa-sel="breadcrumb-user-link"
      elementType={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "propTypes": Object {
            "innerRef": [Function],
            "onClick": [Function],
            "replace": [Function],
            "target": [Function],
            "to": [Function],
          },
          "render": [Function],
        }
      }
      to="/users"
    >
      Users
    </Breadcrumb>
    <Breadcrumb
      elementType={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "propTypes": Object {
            "innerRef": [Function],
            "onClick": [Function],
            "replace": [Function],
            "target": [Function],
            "to": [Function],
          },
          "render": [Function],
        }
      }
      to="/users/1"
    >
      Kristin Patrick
    </Breadcrumb>
    <Breadcrumb
      elementType={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "propTypes": Object {
            "innerRef": [Function],
            "onClick": [Function],
            "replace": [Function],
            "target": [Function],
            "to": [Function],
          },
          "render": [Function],
        }
      }
      to="/users/1/services"
    >
      Select service
    </Breadcrumb>
    <Breadcrumb>
      Select environment
    </Breadcrumb>
  </Breadcrumbs>
  <PageHeader
    heading="EPPI Reviewer"
    preheading="Select environment for"
  />
  <Grid
    elementType="div"
    equalHeight={false}
    gutter="standard"
    horizontalAlignment="left"
    reverse={false}
    verticalAlignment="top"
  >
    <GridItem
      cols={8}
    >
      <StackedNav
        aria-label="Environments"
      >
        <StackedNavLink
          destination="/1/roles"
          elementType={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "propTypes": Object {
                "innerRef": [Function],
                "onClick": [Function],
                "replace": [Function],
                "target": [Function],
                "to": [Function],
              },
              "render": [Function],
            }
          }
        >
          dev
        </StackedNavLink>
        <StackedNavLink
          destination="/2/roles"
          elementType={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "propTypes": Object {
                "innerRef": [Function],
                "onClick": [Function],
                "replace": [Function],
                "target": [Function],
                "to": [Function],
              },
              "render": [Function],
            }
          }
        >
          test
        </StackedNavLink>
      </StackedNav>
    </GridItem>
  </Grid>
</Fragment>
`;
